ls1intum/Artemis
# Feature/Programming-Exercises/Visualize build process for students

### Checklist

### Motivation and Context



Currently in the Artemis client, there is no feedback that a build for a programming exercise is running.
However we do have the information about a running build as there is a submission without a linked result (the submission is created after commit, the result after the build is finished).

### Description


A new service provides information about pending submissions in the client:
- Initially checks for a pending submission via http call
- After that listens on the ws for new submissions
- If there is a pending submission, a spinning circle is shown in the updating-result comp
- When no result is created within 1 minute, an error message will be displayed (see screenshots)

The last pending submission is defined by the following criteria:
- Is the newest submission for a participation based on submissionDate
- Does not have a result yet
- Is not older than 1 minute

:information_source: Known bug: **When a test repository commit is made:** :information_source: 
- No submission is created for the template & solution participation
- The submissions created for the student participations can't be connected to the incoming results as for the submission creation the test repo commitHash is used and for the result the student assignment repo commit is used.
Both issues are fixed in: https://github.com/ls1intum/Artemis/pull/695

**This feature works on all pages that show the updating-result component!**

- [x] Notify the client when the Artemis Server retrieves a new submission over websocket "/topic/participation/" + participation.getId() + "/newSubmission"
- [x] Handle this websocket notification on the client in `updating-result.component.ts`.
- [x] Display an animation with e.g. "Building and Testing" in the result `updating-result.component.html` or `result.component.html` instead of showing the last result. Typically this animation will be shown around 5s.
- [x] When the client finally receives the new result over websocket "/topic/participation/" + participation.getId() + "/newResults", the animation is stopped and the new result is displayed.
- [x] When the client does not receive a new result in less than e.g. 30s or _**60s**_, the client should display an error message that the result was not received and that the user should push some changes again (in the future, we can work on this scenario and try to further improve it)
- [x] Include jhi-alert in course-exercise page so that error can be shown.
- [x] Bug: Buttons in code-editor need to be renabled when the result did not come in in time.

### Steps for Testing


1. Log in to ArTEMiS
2. Open the code-editor
3. Submit code
=> Check that the building process is visualized

### Screenshots



Until a result is returned OR the timer runs out:

![image](https://user-images.githubusercontent.com/28230611/62458654-c4ad5680-b77d-11e9-9a5b-eaaae6f1312d.png)

When the result timer runs out (after 1 min):

![image](https://user-images.githubusercontent.com/28230611/62458777-12c25a00-b77e-11e9-8e93-193bb0de38be.png)

In other areas as e.g. the course-exercise-list:

![image](https://user-images.githubusercontent.com/28230611/62460447-574ff480-b782-11e9-9774-22783b50659c.png)
